File: tests/src/com.google.protobuf/protobuf-javalite/3.25.3/src/test/java/com_google_protobuf/protobuf_javalite/UnsafeUtilTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_google_protobuf.protobuf_javalite;

import static java.lang.invoke.MethodType.methodType;
import static org.assertj.core.api.Assertions.assertThat;

import com.google.protobuf.ByteString;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import org.junit.jupiter.api.Test;

public class UnsafeUtilTest {
    private static final String UNSAFE_UTIL_CLASS_NAME = "com.google.protobuf.UnsafeUtil";

    @Test
    void initializesAndroidMemoryChecksAndAllocatesInstanceWithoutConstructor() throws Throwable {
        Class<?> unsafeUtilClass = Class.forName(
                UNSAFE_UTIL_CLASS_NAME,
                true,
                ByteString.class.getClassLoader()
        );
        MethodHandles.Lookup lookup = MethodHandles.privateLookupIn(
                unsafeUtilClass,
                MethodHandles.lookup()
        );
        MethodHandle allocateInstance = lookup.findStatic(
                unsafeUtilClass,
                "allocateInstance",
                methodType(Object.class, Class.class)
        );

        ConstructorGuardedType instance = (ConstructorGuardedType) allocateInstance.invoke(
                ConstructorGuardedType.class
        );

        assertThat(instance.constructorValue()).isZero();
    }

    public static final class ConstructorGuardedType {
        private final int constructorValue;

        private ConstructorGuardedType() {
            constructorValue = 42;
            throw new AssertionError("Unsafe allocation must not invoke constructors");
        }

        int constructorValue() {
            return constructorValue;
        }
    }
}

File: tests/src/com.typesafe.akka/akka-protobuf-v3_2.13/2.8.8/src/test/scala/com_typesafe_akka/akka_protobuf_v3_2_13/UnsafeUtilTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package libcore.io {
  class Memory {
    def peekLong(address: Long, swap: Boolean): Long = 0L

    def peekLong(address: Int, swap: Boolean): Long = 0L

    def pokeLong(address: Long, value: Long, swap: Boolean): Unit = ()

    def pokeLong(address: Int, value: Long, swap: Boolean): Unit = ()

    def pokeInt(address: Long, value: Int, swap: Boolean): Unit = ()

    def pokeInt(address: Int, value: Int, swap: Boolean): Unit = ()

    def peekInt(address: Long, swap: Boolean): Int = 0

    def peekInt(address: Int, swap: Boolean): Int = 0

    def pokeByte(address: Long, value: Byte): Unit = ()

    def pokeByte(address: Int, value: Byte): Unit = ()

    def peekByte(address: Long): Byte = 0.toByte

    def peekByte(address: Int): Byte = 0.toByte

    def pokeByteArray(address: Long, bytes: Array[Byte], offset: Int, count: Int): Unit = ()

    def pokeByteArray(address: Int, bytes: Array[Byte], offset: Int, count: Int): Unit = ()

    def peekByteArray(address: Long, bytes: Array[Byte], offset: Int, count: Int): Unit = ()

    def peekByteArray(address: Int, bytes: Array[Byte], offset: Int, count: Int): Unit = ()
  }
}

package com_typesafe_akka.akka_protobuf_v3_2_13 {
  import java.lang.invoke.MethodHandles
  import java.lang.invoke.MethodType.methodType

  import org.junit.jupiter.api.Assertions.assertEquals

  import akka.protobufv3.internal.ByteString
  import org.junit.jupiter.api.Test

  class UnsafeUtilTest {
    @Test
    def allocateInstanceInitializesUnsafeUtilAndBypassesConstructors(): Unit = {
      assertEquals("libcore.io.Memory", classOf[libcore.io.Memory].getName)

      val unsafeUtilClass: Class[_] = Class.forName(
        "akka.protobufv3.internal.UnsafeUtil",
        true,
        classOf[ByteString].getClassLoader
      )
      val lookup: MethodHandles.Lookup = MethodHandles.privateLookupIn(
        unsafeUtilClass,
        MethodHandles.lookup()
      )
      val allocateInstance = lookup.findStatic(
        unsafeUtilClass,
        "allocateInstance",
        methodType(classOf[Object], classOf[Class[_]])
      )

      val instance = allocateInstance.invokeWithArguments(classOf[ConstructorGuardedType]).asInstanceOf[ConstructorGuardedType]

      assertEquals(0, instance.constructorValue)
    }
  }

  final class ConstructorGuardedType private () {
    private val initializedValue: Int = 42

    throw new AssertionError("Unsafe allocation must not invoke constructors")

    def constructorValue: Int = initializedValue
  }
}
